Added .NET Framework 4.6.1 back to supported targets

josephmyers opened the pull request:
This provides consuming projects another option if the .NET Standard version is causing issues.

josephmyers committed:
Added .NET Framework 4.6.1 back to supported targets

This provides consuming projects another option if the .NET Standard version is causing issues.

josephmyers committed:
Cleaned up references in LCModel and LCModel.Core

This must have been done for me by the .NET Migration Tool, which I was using at the time. There's no need to be referencing the beta/preview versions of these packages. The solution builds just fine on stable versions. Note I did upgrade "icu.net" to use the star pattern in LCModel.Core, including beta versions. This was required to avoid a binding redirect (I think) in the app.config. Using the original 2.7.1 version caused the binding redirect to point to a version that wasn't available (at least for me), which was causing the unit tests to all throw runtime exceptions.

josephmyers committed:
Referencing .NET Standard version of SharpZipLib

josephmyers committed:
Replaced custom version of structuremap with latest version of original.

This allows the .NET Standard to be used as appropriate. It does require added package reference StrongNamer, but this seems better (more maintainable) than custom making our own version of needed packages. Plus it allows us to follow the same simple pattern for other unsigned packages, including easily changing versions without having to resign manually.

Apparently there is one other small difference between the .patched version and the original (https://github.com/ermshiperete/structuremap), but LCM doesn't seem to be using it.

josephmyers committed:
Revert "Replaced custom version of structuremap with latest version of original."

This reverts commit 6a83c2eb4125243acfd10c66d1b57f0eeae424df.

josephmyers committed:
Referencing latest StructureMap instead of .patched version.

Also referencing StrongNameSigner, since StructureMap author hated signing. This did require changing some constructors for internal classes from internal to public, but within an internal class, these two have the same accessibility, and using public allows StructureMap to find them. The one "public class / internal constructor" class was Virtuals.cs. If it's a big deal to expose this constructor, we can definitely fix it.

The advantages to avoiding manually creating custom code should be obvious: lower maintenance, future version support (though StructureMap is retired), and with StrongNameSigner the ability to apply one solution for all (unsigned) packages.

ermshiperete commented on src/SIL.LCModel/Application/Impl/DomainDataByFlid.cs:
```diff
@@ -42,7 +42,7 @@ internal sealed class DomainDataByFlid : ISilDataAccessManaged
 		/// CmObject identity can only be guaranteed by using their Guids (or using '==' in code).
 		/// </remarks>
-		internal DomainDataByFlid(ICmObjectRepository cmObjectRepository, IStTextRepository stTxtRepository,
+		public DomainDataByFlid(ICmObjectRepository cmObjectRepository, IStTextRepository stTxtRepository,
```
I'm not too keen on having these c'tors public even though it doesn't change anything in regards to the binary.

The reason is that all or at least most of these classes are implementation details and never intended to be created from outside of liblcm. Changing the c'tor to public muddles this intention a bit.

ermshiperete commented on src/SIL.LCModel/LcmCache.cs:
```diff
@@ -1434,7 +1434,7 @@ internal IFwMetaDataCache MetaDataCache
 		/// Constructor.
 		/// </summary>
-		internal LcmCache()
+		public LcmCache()
```
This can't be `public` - `LcmCache` isn't intended to be created directly, only through one of the `CreateCache...` methods.

ermshiperete reviewed (changes requested):
I didn't do a full review, but have a few comments:

Using StrongNameSigner seems like a good idea (I originally went down that route as well) - unfortunately it causes problems in the projects that use the liblcm nuget package: StructureMap comes in as a transitive dependency, but is unsigned. This means all projects using liblcm directly or indirectly would need to reference StrongNameSigner as well. And what's more serious is that StrongNameSigner generates a signing key on the fly, so every machine creates a slightly different signed assembly. And because the liblcm nuget package/assembly mentions the key in the StructureMap reference, it won't work on any other machine.

josephmyers committed:
Revert

josephmyers commented:
Thanks for the look -- it sounds like that's the problem Hasso was having. I reverted that change and will get back on this tomorrow

josephmyers committed:
Using v10 for libpalaso packages

This was necessary only to use the .NET Standard versions of SIL.Lexicon and SIL.TestUtilities. v9 has it for the other packages

josephmyers commented:
In the interest of stabilizing this branch for potential integration testing, I can stop pushes here. It will block me on the structuremap.patched work, but at this point I've realized that that effort is high-risk
